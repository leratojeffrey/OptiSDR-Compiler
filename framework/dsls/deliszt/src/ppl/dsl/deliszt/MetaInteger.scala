package ppl.dsl.deliszt

/**
 * author: Michael Wu (mikemwu@stanford.edu)
 * last modified: 04/28/2011
 *
 * Pervasive Parallelism Laboratory (PPL)
 * Stanford University
 */

/*
 * MetaInteger.scala
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

abstract class IntM {
  type Suc <: IntM
  type Add[N <: IntM] <: IntM
}

object EnsureSize {
  implicit def EnsureSizeZero[T <: IntM] =
  new EnsureSize[Zero,Succ[T]] {
    def get(n : Int) : Int = n
  }

  implicit def EnsureSizeSucc[T <: IntM, TT <: IntM](implicit fn : EnsureSize[TT,T]) =
  new EnsureSize[Succ[TT],Succ[T]] {
    def get(n : Int) : Int = fn.get(n+1)
  }
}

//meta-function(?)(what is the term for this thing?)
//if IDX is less than SIZE, then an implicit object of type EnsureSize[IDX,SIZE] will
//exist with a idx function that returns the number IDX as an integer
abstract class EnsureSize[TT <: IntM,T <: IntM] {
  def get(n : Int) : Int
  def idx = get(0)
}

object MVal {
  implicit val MValZero = new MVal[Zero] {
    val value = 0
  }

  implicit def MValSucc[T <: IntM](implicit f: MVal[T]) = new MVal[Succ[T]] {
    val value = f.value + 1
  }
}

abstract class MVal[T] {
  val value: Int
}

class Zero extends IntM {
  type Suc = Succ[Zero]
  type Add[N <: IntM] = N
}

class Succ[N <: IntM] extends IntM {
  type Suc = Succ[Succ[N]]
  type Add[M <: IntM] = (N#Add[M])#Suc
}

object MetaInteger {
  type +[N1 <: IntM, N2 <: IntM] = N1#Add[N2]

  //Scala doesn't provide meta-integers like C++'s typename<int N>
  //so we have to define some integers here
  //this is a simplification of:
  //http://michid.wordpress.com/code/meta-programming-with-scala-part-i-addition/
  //somelike like Church encoding but the numbers are themselves objects and not
  //functions for simplicity since we don't need to define addition on them
  //as far I can tell ... if necessary I suppose we can define multiplcation
  //on these if we needed numbers higher than 9
  type _0 = Zero
  type _1 = Succ[_0]
  type _2 = Succ[_1]
  type _3 = Succ[_2]
  type _4 = Succ[_3]
  type _5 = Succ[_4]
  type _6 = Succ[_5]
  type _7 = Succ[_6]
  type _8 = Succ[_7]
  type _9 = Succ[_8]
  type _10 = Succ[_9]
  type _11 = Succ[_10]
  type _12 = Succ[_11]
  type _13 = Succ[_12]
  type _14 = Succ[_13]
  type _15 = Succ[_14]
  type _16 = Succ[_15]
  type _17 = Succ[_16]
  type _18 = Succ[_17]
  type _19 = Succ[_18]
  type _20 = Succ[_19]
  type _21 = Succ[_20]
  type _22 = Succ[_21]
  
  //for convience we provide objects representing each number as well
  //so you can avoid passing parameter arguments
  val _0 : _0 = new _0
  val _1 : _1 = new _1
  val _2 : _2 = new _2
  val _3 : _3 = new _3
  val _4 : _4 = new _4
  val _5 : _5 = new _5
  val _6 : _6 = new _6
  val _7 : _7 = new _7
  val _8 : _8 = new _8
  val _9 : _9 = new _9
  val _10 : _10 = new _10
  val _11 : _11 = new _11
  val _12 : _12 = new _12
  val _13 : _13 = new _13
  val _14 : _14 = new _14
  val _15 : _15 = new _15
  val _16 : _16 = new _16
  val _17 : _17 = new _17
  val _18 : _18 = new _18
  val _19 : _19 = new _19
  val _20 : _20 = new _20
  val _21 : _21 = new _21
  val _22 : _22 = new _22

  def MIntDepth[T <: IntM]()(implicit n: MVal[T]) = n.value
  def MIntDepth[T <: IntM](m : T)(implicit n: MVal[T]) = n.value
}
