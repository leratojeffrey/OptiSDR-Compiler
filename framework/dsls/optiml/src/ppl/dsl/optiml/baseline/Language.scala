package ppl.dsl.optiml.baseline

object LanguageDefs {
  val INITIAL_SEED = 100
  var randRef = new scala.util.Random(INITIAL_SEED)
  var _time = 0L
  
  def random(): Double = randRef.nextDouble
    
  def abs(d: Double) = implicitly[Numeric[Double]].abs(d) //java.lang.Math.abs(d)
  
  def sum[B:Numeric:Manifest](st: Int, en: Int)(f: Int => B) = {
    var acc = f(st)
    for (i <- st+1 until en) {
      acc = implicitly[Numeric[B]].plus(acc,f(i))
    }
    acc
  }
  
  def tic() {
    _time = System.currentTimeMillis()
  }
  
  def toc() {
    val now = System.currentTimeMillis()
    println("elapsed: " + (now-_time)/1000.0)
  }
}