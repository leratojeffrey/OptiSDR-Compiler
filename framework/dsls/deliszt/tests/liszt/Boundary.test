import Liszt.Language._
import Liszt.MetaInteger._

@lisztcode
object Boundary {
    val inlet = BoundarySet[Face]("inlet") ;
    val outlet = BoundarySet[Face]("outlet") ;
    val farField = BoundarySet[Face]("far_field") ;
    val fluid = BoundarySet[Cell]("fluid") ;
    val defaultInterior = BoundarySet[Face]("default-interior") ;
    val interior = BoundarySet[Face]("type:interior") ;
    val bound1 = BoundarySet[Face]("bound1") ;
    val bound2 = BoundarySet[Face]("bound2") ;
    val bound3 = BoundarySet[Face]("bound3") ;
    val bound4 = BoundarySet[Cell]("bound4") ;
    val bound5 = BoundarySet[Face]("bound5") ;

    def main() {
        var inlet_count = 0;
        var outlet_count = 0;
        var farField_count = 0;
        var fluid_count = 0;
        var interior_count = 0;
        var defaultInterior_count = 0;
        var bound1_count = 0;
        var bound2_count = 0;
        var bound3_count = 0;
        var bound4_count = 0;
        var bound5_count = 0;
        for ( f <- inlet ) {
            if ( !( ID(inside(f)) == 0) ) {
                if ( !(ID(outside(f)) == 0 ) ) {
                    Print( "error, not boundary face: ", ID(f) ) ;
                }
            } 
            inlet_count += 1;
        }
        for ( f <- outlet ) {
            if ( !( ID(inside(f)) == 0) ) {
                                if ( !(ID(outside(f)) == 0 ) ) {
                                        Print( "error, not boundary face: ", ID(f) ) ;
                                }
                        }
            outlet_count += 1;
                }
        for ( f <- farField ) {
            if ( !( ID(inside(f)) == 0) ) {
                                if ( !(ID(outside(f)) == 0 ) ) {
                                        Print( "error, not boundary face: ", ID(f) ) ;
                                }
                        }
            farField_count += 1
                }
        for ( f <- defaultInterior ) { defaultInterior_count += 1 }
        for ( f <- interior ) { interior_count += 1 }
        for ( f <- bound1 ) { bound1_count += 1 }
        for ( f <- bound2 ) { bound2_count += 1 }
        for ( f <- bound3 ) { bound3_count += 1 }
        for ( c <- bound4 ) { bound4_count += 1 }
        for ( c <- bound5 ) { bound5_count += 1 }
        for ( c <- fluid ) { fluid_count += 1 }
        if (fluid_count != bound4_count) {
            Print( "error, bound4 and fluid counts mismatch. Expected ", fluid_count, ", got ", bound4_count )
        }
        Print( "inlet: ", inlet_count )
        Print( "outlet: ", outlet_count )
        Print( "farField: ", farField_count )
        Print( "fluid: ", fluid_count )
        Print( "default-interior: ", defaultInterior_count )
        Print( "type:interior: ", interior_count )
        Print( "bound1: ", bound1_count )
        Print( "bound2: ", bound2_count )
        Print( "bound3: ", bound3_count )
        Print( "bound4: ", bound4_count )
        Print( "bound5: ", bound5_count )
    }
}
