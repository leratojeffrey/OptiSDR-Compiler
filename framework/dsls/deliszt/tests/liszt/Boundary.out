11 liszt: inlet: 25
12 liszt: outlet: 25
13 liszt: farField: 200
14 liszt: fluid: 249
15 liszt: default-interior: 625
16 liszt: type:interior: 625
17 liszt: bound1: 50
18 liszt: bound2: 50
19 liszt: bound3: 875
20 liszt: bound4: 249
21 liszt: bound5: 20
